core/appeal/service.go

```diff
@@ -15,72 +25,105 @@ import (
 	"github.com/odpf/salt/log"
 )
 
+var (
+	AuditKeyBulkInsert = "appeal.bulkInsert"
+	AuditKeyCancel     = "appeal.cancel"
+	AuditKeyApprove    = "appeal.approve"
+	AuditKeyReject     = "appeal.reject"
+	AuditKeyRevoke     = "appeal.revoke"
+	AuditKeyExtend     = "appeal.extend"
+)
+
 var TimeNow = time.Now
 
+type repository interface {
+	store.AppealRepository
+}
+
+type iamManager interface {
+	domain.IAMManager
+}
+
+type notifier interface {
+	notifiers.Client
+}
+
 type policyService interface {
-	Find() ([]*domain.Policy, error)
-	GetOne(string, uint) (*domain.Policy, error)
+	Find(context.Context) ([]*domain.Policy, error)
+	GetOne(context.Context, string, uint) (*domain.Policy, error)
 }
 
 type approvalService interface {
-	AdvanceApproval(*domain.Appeal) error
+	AdvanceApproval(context.Context, *domain.Appeal) error
 }
 
 type providerService interface {
-	Find() ([]*domain.Provider, error)
-	GrantAccess(*domain.Appeal) error
-	RevokeAccess(*domain.Appeal) error
-	ValidateAppeal(*domain.Appeal, *domain.Provider) error
+	Find(context.Context) ([]*domain.Provider, error)
+	GrantAccess(context.Context, *domain.Appeal) error
+	RevokeAccess(context.Context, *domain.Appeal) error
+	ValidateAppeal(context.Context, *domain.Appeal, *domain.Provider) error
 }
 
 type resourceService interface {
-	Find(map[string]interface{}) ([]*domain.Resource, error)
-	Get(*domain.ResourceIdentifier) (*domain.Resource, error)
+	Find(context.Context, map[string]interface{}) ([]*domain.Resource, error)
+	Get(context.Context, *domain.ResourceIdentifier) (*domain.Resource, error)
+}
+
+type auditLogger interface {
+	Log(ctx context.Context, action string, data interface{}) error
+}
+
+type ServiceOptions struct {
+	Repository      repository
+	ApprovalService approvalService
+	ResourceService resourceService
+	ProviderService providerService
+	PolicyService   policyService
+	IAMManager      iamManager
+
+	Notifier    notifier
+	Validator   *validator.Validate
+	Logger      log.Logger
+	AuditLogger auditLogger
 }
 
 // Service handling the business logics
 type Service struct {
-	repo store.AppealRepository
-
+	repo            repository
 	approvalService approvalService
 	resourceService resourceService
 	providerService providerService
 	policyService   policyService
 	iam             domain.IAMManager
-	notifier        notifiers.Client
-	logger          log.Logger
 
-	validator *validator.Validate
-	TimeNow   func() time.Time
+	notifier    notifier
+	validator   *validator.Validate
+	logger      log.Logger
+	auditLogger auditLogger
+
+	TimeNow func() time.Time
 }
 
 // NewService returns service struct
-func NewService(
-	appealRepository store.AppealRepository,
-	approvalService approvalService,
-	resourceService resourceService,
-	providerService providerService,
-	policyService policyService,
-	iam domain.IAMManager,
-	notifier notifiers.Client,
-	logger log.Logger,
-) *Service {
+func NewService(opts ServiceOptions) *Service {
```

Comment:
If we are using options here, should we use it like [golang options](https://www.google.com/search?q=golang+function+options&oq=golang+function+options&aqs=chrome..69i57.2961j0j7&sourceid=chrome&ie=UTF-8) behave? The fields are optional and has default value if not passed.

Reply:
just learned that this should've been dependencies instead of options, changing the struct name to `ServiceDeps` will be more reasonable now in my opinion